contracts/EscrowAndStaking/CrossChainCollector.sol

```diff
@@ -0,0 +1,408 @@
+        if (numOfMessagesReceived > 0) {
+            chainSelector_ = messagesReceivedFromChainSelector[numOfMessagesReceived - 1];
+        }
+    }
+
+    /* ========== EXTERNAL FUNCTIONS ========== */
+
+    function sendOnClosePeriod(
+        uint _totalStakedLastPeriodEnd,
+        uint _totalEscrowedLastPeriodEnd,
+        uint _bonusPoints,
+        uint _revShare
+    ) external {
+        require(msg.sender == stakingThales, "InvSender");
+        if (masterCollector == address(this)) {
+            _storeRewards(
+                masterCollectorChain,
+                _totalStakedLastPeriodEnd,
+                _totalEscrowedLastPeriodEnd,
+                _bonusPoints,
+                _revShare
+            );
+            ++numOfCollectedResultsForThisPeriod;
+            if (numOfCollectedResultsForThisPeriod == numOfActiveCollectors) {
+                readyToBroadcast = true;
+            }
+        } else {
+            bytes memory message = abi.encode(
+                _totalStakedLastPeriodEnd,
+                _totalEscrowedLastPeriodEnd,
+                _bonusPoints,
+                _revShare
+            );
+            _sendMessageToChain(masterCollectorChain, message);
+        }
+        emit SentOnClosePeriod(_totalStakedLastPeriodEnd, _totalEscrowedLastPeriodEnd, _bonusPoints, _revShare);
+    }
+
+    function broadcastMessageToAll() external onlyOwner {
+        if (readyToBroadcast) {
+            _broadcastMessageToAll();
+            numOfCollectedResultsForThisPeriod = 0;
+            readyToBroadcast = false;
+            ++period;
+            if (weeklyRewardsDecreaseFactor > 0) {
+                _setRewardsForNextPeriod();
+            }
+        }
+    }
+
+    /* ========== INTERNAL FUNCTIONS ========== */
+
+    /// handle a received message
+    function _ccipReceive(Client.Any2EVMMessage memory any2EvmMessage) internal override {
+        // decoding the message
+        if (!messageIdAlreadyReceived[any2EvmMessage.messageId]) {
+            address sender = abi.decode(any2EvmMessage.sender, (address));
+            messagesReceivedFromChainSelector[numOfMessagesReceived] = any2EvmMessage.sourceChainSelector;
+            messagesReceived[numOfMessagesReceived] = any2EvmMessage.data;
+            messageIdAlreadyReceived[any2EvmMessage.messageId] = true;
+            ++numOfMessagesReceived;
+
+            if (masterCollector == address(this)) {
+                uint sourceChainSelector = any2EvmMessage.sourceChainSelector;
+                if (readOnlyMode) {
+                    if (collectorAddress[sourceChainSelector] == sender) {
+                        _calculateRewards(any2EvmMessage.data, sourceChainSelector);
+                    }
+                    ++numOfCollectedResultsForThisPeriod;
+                    if (numOfCollectedResultsForThisPeriod == numOfActiveCollectors) {
+                        readyToBroadcast = true;
```

Comment:
can be replaced with readyToBroadcast = numOfCollectedResultsForThisPeriod == numOfActiveCollectors)

Reply:
yes, however in `readOnlyMode==true` can happen that `numOfCollectedResultsForThisPeriod >= numOfActiveCollectors`

Reply:
> yes, however in `readOnlyMode==true` can happen that `numOfCollectedResultsForThisPeriod >= numOfActiveCollectors`
> 
> I would say with the if clause only a single store operation is performed in each period, unlike `readyToBroadcast = numOfCollectedResultsForThisPeriod == numOfActiveCollectors`